docs: Update v1 migration guide to enable the conversion webhooks by helm controller (#6703)
aws/karpenter-provider-aws: website/content/en/preview/upgrading/v1-migration.md

2. Review for breaking changes between v0.33 and v0.37: If you are already running Karpenter v0.37.x, you can skip this step. If you are running an earlier Karpenter version, you need to review the [Upgrade Guide]({{<ref "upgrade-guide#upgrading-to-0320" >}}) for each minor release.

> 3. Set environment variables for your cluster to upgrade to the latest patch version of the current Karpenter version you're running on:

Reviewer: Do we need to give more info about why we are telling users to do this? Otherwise, I could see questions about why I wouldn't just go straight from my current version to 1.0.0

Author: I added a note at the beginning of the upgrade guide

    ```bash
    export KARPENTER_NAMESPACE=kube-system
    export KARPENTER_IAM_ROLE_ARN="arn:${AWS_PARTITION}:iam::${AWS_ACCOUNT_ID}:role/${CLUSTER_NAME}-karpenter"
    export AWS_PARTITION="aws" # if you are not using standard partitions, you may need to configure to aws-cn / aws-us-gov
    export CLUSTER_NAME="${USER}-karpenter-demo"
    export AWS_REGION="us-west-2"
    export AWS_ACCOUNT_ID="$(aws sts get-caller-identity --query Account --output text)"
    ```

> 4. Set environment variables for first upgrading to latest patch version of your current minor version's Karpenter controller

Reviewer: ```suggestion
4. Set environment variables for upgrading to the latest patch version:
```

nit: IMO, I find the verbosity a bit confusing

    ```bash
    export KARPENTER_VERSION=<latest patch version of your current v1beta1 minor version>
    ```

> 5. Update your existing policy using the following:

Reviewer: We shouldn't have changed the IAM policy on a patch version so I don't think that it's necessary to ask users to walk through the IAM upgrade multiple times. Did we change things in the IAM policy within a minor version?

> 

Reviewer 2: Let's add more info on the 3. step as an overview for what's happening. 

3. Set environment variables for your cluster to upgrade to the latest patch version of the current Karpenter version you're running on.

    ```bash
    export KARPENTER_NAMESPACE=kube-system
    export KARPENTER_IAM_ROLE_ARN="arn:${AWS_PARTITION}:iam::${AWS_ACCOUNT_ID}:role/${CLUSTER_NAME}-karpenter"
    export AWS_PARTITION="aws" # if you are not using standard partitions, you may need to configure to aws-cn / aws-us-gov
    export CLUSTER_NAME="${USER}-karpenter-demo"
    export AWS_REGION="us-west-2"
    export AWS_ACCOUNT_ID="$(aws sts get-caller-identity --query Account --output text)"
    ```

> 4. Set environment variables for first upgrading to v0.37.1

Reviewer 2: This should probably be a function of whatever minor version they're on, right? We're not requiring users migrate to v0.37 from v0.33.

Author: Yeah, you are right on that one. I will update

        --parameter-overrides "ClusterName=${CLUSTER_NAME}"
    ```

> 6. Apply the v0.37.1 Custom Resource Definitions (CRDs):

Reviewer 2: ```suggestion
6. Apply the latest patch version of your current minor version's Custom Resource Definitions (CRDs):
```

    curl -fsSL https://raw.githubusercontent.com/aws/karpenter-provider-aws/v"${KARPENTER_VERSION}"/website/content/en/docs/getting-started/getting-started-with-karpenter/cloudformation.yaml > ${TEMPOUT} \
        && aws cloudformation deploy \
        --stack-name "Karpenter-${CLUSTER_NAME}" \
        --template-file "${TEMPOUT}" \
        --capabilities CAPABILITY_NAMED_IAM \
        --parameter-overrides "ClusterName=${CLUSTER_NAME}"
    ```

10. Apply the v1.0.0 Custom Resource Definitions (CRDs):

    ```bash
    helm upgrade --install karpenter-crd oci://public.ecr.aws/karpenter/karpenter-crd --version "${KARPENTER_VERSION}" --namespace "${KARPENTER_NAMESPACE}" --create-namespace \
        --set webhook.enabled=true \
        --set webhook.serviceName=karpenter \
        --set webhook.serviceNamespace="${KARPENTER_NAMESPACE}" \
        --set webhook.port=8443
    ```

11. Upgrade Karpenter to the new version. At the end of this step, conversion webhooks run to convert the Karpenter CRDs to v1.

    ```bash
    helm upgrade --install karpenter oci://public.ecr.aws/karpenter/karpenter --version ${KARPENTER_VERSION} --namespace "${KARPENTER_NAMESPACE}" --create-namespace \

>         --set serviceAccount.annotations."eks\.amazonaws\.com/role-arn"=${KARPENTER_IAM_ROLE_ARN} \

Reviewer: Does it make sense to put an addendum next to this line that says that this isn't needed if using EKS pod identity?

Author: I'm not sure it make much of a difference here, but I can add one above the command

   ```bash
   helm upgrade --install karpenter-crd oci://public.ecr.aws/karpenter/karpenter-crd --version "${KARPENTER_VERSION}" --namespace "${KARPENTER_NAMESPACE}" --create-namespace \

>         --set webhook.enabled=false \

Reviewer 2: I thought we were saying that we could just go straight into upgrading to the v1 with webhook enabled?
```suggestion
        --set webhook.enabled=false \
```

Author: Yeah, you can. This is proposing that you can upgrade the CRDs and without creating downtime to the pre-v1 controller.

Reviewer 2: I'm missing the point of why'd we want to propose this? Seems like we'd force them to do more upgrades. We should just recommend the safest shortest path

Author: We tell them to disable the webhooks in the CRD, not in the controller. When they run the helm upgrade for the controller with the webhooks enabled the post install hook will add the conversion block. So no extra steps needed from users

Reviewer: This feels like one-too-many steps here. We shouldn't need to go from disabling webhooks to enabling webhooks like this IMO, seems like an extra step that really isn't necessary to avoid downtime

        --parameter-overrides "ClusterName=${CLUSTER_NAME}"
    ```

10. Apply the v1.0.0 Custom Resource Definitions (CRDs):

    ```bash
    helm upgrade --install karpenter-crd oci://public.ecr.aws/karpenter/karpenter-crd --version "${KARPENTER_VERSION}" --namespace "${KARPENTER_NAMESPACE}" --create-namespace \
        --set webhook.enabled=true \
        --set webhook.serviceName=karpenter \
        --set webhook.serviceNamespace="${KARPENTER_NAMESPACE}" \
        --set webhook.port=8443
    ```

11. Upgrade Karpenter to the new version. At the end of this step, conversion webhooks run to convert the Karpenter CRDs to v1.

    ```bash

>     # Service account annotion can be droped when using pod identity 

Reviewer 2: ```suggestion
    # Service account annotion can be dropped when using pod identity 
```
